pkg/scheduler/controller/scheduler_control.go

```diff
@@ -0,0 +1,375 @@
+	"k8s.io/client-go/tools/record"
+	"k8s.io/client-go/util/retry"
+	"strings"
+	"time"
+
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+
+	gardencorev1alpha1 "github.com/gardener/gardener/pkg/apis/core/v1alpha1"
+	gardencorev1alpha1helper "github.com/gardener/gardener/pkg/apis/core/v1alpha1/helper"
+	gardenv1beta1 "github.com/gardener/gardener/pkg/apis/garden/v1beta1"
+	gardenhelper "github.com/gardener/gardener/pkg/apis/garden/v1beta1/helper"
+	gardeninformers "github.com/gardener/gardener/pkg/client/garden/informers/externalversions"
+	gardenlisters "github.com/gardener/gardener/pkg/client/garden/listers/garden/v1beta1"
+	"github.com/gardener/gardener/pkg/client/kubernetes"
+	"github.com/gardener/gardener/pkg/logger"
+	"github.com/gardener/gardener/pkg/operation/common"
+	"github.com/gardener/gardener/pkg/scheduler/apis/config"
+	schedulerutils "github.com/gardener/gardener/pkg/scheduler/utils"
+	"github.com/gardener/gardener/pkg/utils"
+	kutil "github.com/gardener/gardener/pkg/utils/kubernetes"
+)
+
+func (c *SchedulerController) shootAdd(obj interface{}) {
+	key, err := cache.MetaNamespaceKeyFunc(obj)
+	if err != nil {
+		logger.Logger.Errorf("Couldn't get key for object %+v: %v", obj, err)
+		return
+	}
+	c.shootQueue.Add(key)
+}
+
+func (c *SchedulerController) shootUpdate(oldObj, newObj interface{}) {
+	var (
+		oldShoot      = oldObj.(*gardenv1beta1.Shoot)
+		newShoot      = newObj.(*gardenv1beta1.Shoot)
+		specChanged   = !apiequality.Semantic.DeepEqual(oldShoot.Spec, newShoot.Spec)
+		statusChanged = !apiequality.Semantic.DeepEqual(oldShoot.Status, newShoot.Status)
+	)
+
+	if !specChanged && statusChanged {
+		return
+	}
+
+	// do not schedule when annotated for deletion
+	if isAnnotatedForDeletion(newShoot) && !isAnnotatedForDeletion(oldShoot) {
+		logger.Logger.Infof("Ignoring shoot '%s' because it has been marked for deletion", newShoot.Name)
+		return
+	} else if isAnnotatedForDeletion(newShoot) {
+		return
+	}
+
+	if oldShoot.Spec.Cloud.Seed == newShoot.Spec.Cloud.Seed {
+		logger.Logger.Infof("Ignoring update to shoot '%s' because its seed has not been modified", newShoot.Name)
+		return
+	}
+
+	c.shootAdd(newObj)
+}
+
+func (c *SchedulerController) reconcileShootKey(ctx context.Context, key string) error {
+	namespace, name, err := cache.SplitMetaNamespaceKey(key)
+	if err != nil {
+		return err
+	}
+
+	shoot, err := c.shootLister.Shoots(namespace).Get(name)
+	if apierrors.IsNotFound(err) {
+		logger.Logger.Debugf("[SCHEDULER SHOOT RECONCILE] %s - skipping because Shoot has been deleted", key)
+		return nil
+	}
+	if err != nil {
+		logger.Logger.Infof("[SCHEDULER SHOOT RECONCILE] %s - unable to retrieve object from store: %v", key, err)
+		return err
+	}
+
+	if err := c.control.ScheduleShoot(ctx, shoot, key); err != nil {
+		logger.Logger.Infof("Failed to schedule Shoot '%s': %+v", shoot.Name, err)
+		willRetry := !utils.TimeElapsed(shoot.Status.RetryCycleStartTime, c.config.RetryDuration.Duration)
```

Comment:
Why are you having this complicated logic? The retry cycle start time is only meaningful for the ctrl-mgr, I guess. Can you just return the error of `ScheduleShoot` here?

Reply:
The scheduler should stop to try to reschedule a shoot after a certain amount of time (24 h default). 
I guess its safe to assume that a shoot is unschedulable after 24 hours of retries.

When its scheduled, the RetryCycleTime is set to nil again.

Reply:
Can't we simply make use of the exponential backoff mechanism of the workqueue itself? Why building some own logic here for this? We don't make expensive external calls, so using the workqueue should be fine, or?